packages/cli/src/ui/components/MainContent.tsx

```diff
@@ -32,65 +36,116 @@ export const MainContent = () => {
     availableTerminalHeight,
   } = uiState;
 
-  const historyItems = [
-    <AppHeader key="app-header" version={version} />,
-    ...uiState.history.map((h) => (
-      <HistoryItemDisplay
-        terminalWidth={mainAreaWidth}
-        availableTerminalHeight={staticAreaMaxItemHeight}
-        availableTerminalHeightGemini={MAX_GEMINI_MESSAGE_LINES}
-        key={h.id}
-        item={h}
-        isPending={false}
-        commands={uiState.slashCommands}
-      />
-    )),
-  ];
+  const historyItems = uiState.history.map((h) => (
+    <HistoryItemDisplay
+      terminalWidth={mainAreaWidth}
+      availableTerminalHeight={staticAreaMaxItemHeight}
+      availableTerminalHeightGemini={MAX_GEMINI_MESSAGE_LINES}
+      key={h.id}
+      item={h}
+      isPending={false}
+      commands={uiState.slashCommands}
+    />
+  ));
+
+  const pendingItems = useMemo(
+    () => (
+      <OverflowProvider>
+        <Box flexDirection="column">
+          {pendingHistoryItems.map((item, i) => (
+            <HistoryItemDisplay
+              key={i}
+              availableTerminalHeight={
+                uiState.constrainHeight ? availableTerminalHeight : undefined
+              }
+              terminalWidth={mainAreaWidth}
+              item={{ ...item, id: 0 }}
+              isPending={true}
+              isFocused={!uiState.isEditorDialogOpen}
+              activeShellPtyId={uiState.activePtyId}
+              embeddedShellFocused={uiState.embeddedShellFocused}
+            />
+          ))}
+          <ShowMoreLines constrainHeight={uiState.constrainHeight} />
+        </Box>
+      </OverflowProvider>
+    ),
+    [
+      pendingHistoryItems,
+      uiState.constrainHeight,
+      availableTerminalHeight,
+      mainAreaWidth,
+      uiState.isEditorDialogOpen,
+      uiState.activePtyId,
+      uiState.embeddedShellFocused,
+    ],
+  );
 
-  const pendingItems = (
-    <OverflowProvider>
-      <Box flexDirection="column" width={mainAreaWidth}>
-        {pendingHistoryItems.map((item, i) => (
-          <HistoryItemDisplay
-            key={i}
-            availableTerminalHeight={
-              uiState.constrainHeight ? availableTerminalHeight : undefined
-            }
+  const virtualizedData = useMemo(
+    () => [
+      { type: 'header' as const },
+      ...uiState.history.map((item) => ({ type: 'history' as const, item })),
+      { type: 'pending' as const },
+    ],
+    [uiState.history],
+  );
+
+  const renderItem = useCallback(
+    ({ item }: { item: (typeof virtualizedData)[number] }) => {
+      if (item.type === 'header') {
+        return <MemoizedAppHeader key="app-header" version={version} />;
+      } else if (item.type === 'history') {
+        return (
+          <MemoizedHistoryItemDisplay
```

Comment:
I memoized a few things here to improve performance. Previously it didn't matter as these items were mainly rendered in static.